pkgs/zitadel: buf resiliency & update script

benaryorg opened the pull request:
A somewhat helpful update script which also changes how we pull in the *buf.build* dependencies, hopefully causing less breakage whenever *buf* receives an update (see #388353).

Note: the script can be used to pull the package from 2.71.1 to 2.71.3 and it seems to work?
~~Also checking the GitHub API should be done better considering zitadel supports multiple versions as far as I can tell?~~ *Solved in the current version.*

In any case, the code quality and especially readability is terrible, this should be a starting point for improvement.

Below the usual checklist is a write-up on the current state of covering *buf* in nixpkgs, in relation to the zitadel package which was intended for a comment in a PR originally, however since this commit implements one of the approaches we may as well discuss it as part of the PR.
Please note that I am not an expert in zitadel, protobuf, golang, Nix, and especially not buf.build or buf in general, so many grains of salt adviced.

# Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

---

This solution has opted for *pointing to buf export data*, which requires that a bunch of hashes from FODs get precomputed.
Because the FOD is a derivation (three of them actually) with a command that actually runs, this is not trivial.
However the script also covers the repo itself by checking GitHub for updates, yarn, the protobuf dependencies, and since it is the only FOD hash left also the go modules.
The repo and yarn work with the usual pre-fetchers.
Protobuf by running the buf commands with modified cache dir (so as to not bother anyone with extra files lying around).
And the go modules, by virtue of being the only thing left, can be resolved by trying to build the package with lib.fakeHash and getting the FOD error and parsing that.

# buf building

Let's start by describing the problem, our prior approach, and its issues.
The problem is relatively simple, we need to build the protobuf specs and get bindings for them, however those specs depend on other specs, and the only reference we have is to an online service which gives us artifacts.
So since we don't have the source repositories for any of this we have to make do by fetching the artifacts in an FOD (to allow us network access for fetching it) using upstream tooling.
The approach has the issue of the FOD being dependent on the upstream tooling which on version change occasionally changes its output and therefore the FOD hash.

Based on this a change can mean several possible things, which can roughly be classified by how the build is changed.
Since many of those approaches have obstacles that are the same for each subset there is a context section further down so I don't have to explain the same thing multiple times.
Terms described in the context have *emphasis*.

## same build, extending the fetch

### fetchtarball

Since the dependency is based on the *BSR commit* we could include that commit in a fetcher much like we include git revs.
If we had a deterministic fetcher which could fetch the *BSR commit* (or rather its output), then we would not depend on *buf*, and therefore not be affected by its version changes as part of the FOD.
The build process would then be separate but offline, thus not an FOD, so a version change in *buf* would also cause a rebuild of the bindings and package.

However *buf.build* has apparently [disabled (or broken) their tar export with the examples on their docs not working](https://buf.build/docs/bsr/module/export/#export-with-curl) as far as I can tell.
This means we cannot use a tarball fetcher.

### BUF_CACHE_DIR

By keeping the build exactly the same but running part of it (`buf dep graph`) beforehand to fetch all dependencies, FODing the cache, and reinjecting it into the build, we can remove the FOD from the actual build.
However testing *nixos-24.11* vs. *master* nixpkgs *buf*, the cache dir shows differences (newer version creates an empty folder for plugins), which means we're back to the point where we depend on *buf* versions for the output of an *FOD*, which on *buf* updates can cause hash mismatches.
The only difference is that now we're hashing internals instead of output-related data, which might prove more stable, but can also change without notice (e.g. in a patch version) so even small updates can cause this.
Prebuilding the cache dir ourselves doesn't seem quite feasible since it includes a plethora of extra files including YAML files generated from build metadata.

### naive buf export

*buf* has an export subcommand which fetches the build artifacts and outputs them in an output directory.
Since this includes only the built output of what is part of the secure hash of the *BSR commit*, it will most likely not change in between *buf* versions.
The problem here lies with integrating it into the build since we cannot point *buf* at the directory and say "go" without modifying the build files significantly.

## changing the build

### pointing to buf export data

If we used the naive *buf* export approach from above, we'd have to somehow point *buf* at the new data.
Unfortunately the *buf.yaml* is structured in a way that we cannot easily replace those paths since zitadel uses the v1 version and the "deps" key which only accepts a *BSR* reference and nothing else (no path, no git repo, no nothing).
Moving the import of that data elsewhere would require changes to the yaml files which is quite hacky tbh.
Especially since we're now pulling in already built protobuf files vs. the original source repositories.
Luckily protobuf itself has an import system which accomodates files lying around in directories, that's how it usually works without *buf*.
By simply copying the prefetched directories over and erasing the deps from the *buf.yaml* as well as *buf.lock* and instead adding them to the buf.work.yaml which lists the project-wide directories to be built we can shorten out every *buf* logic pertaining to those files and treat them as natural protobuf files.
It still involves wild hacking around in *buf* configuration, however the code pulls from the repository itself.
As long as upstream doesn't introduce any indirect dependencies this should continue to work indefinitely.

## source code all the way down

Instead of fetching the build artifacts, which are the thing fetched online, we could try and fetch the sources of the dependencies.
This has multiple issues explored in the next few sections.

### getting the source repository

First of all we would have to get the source repositories, in case upstream ever changes these.
These are technically easily guessable (replace *buf.build* with github.com works for most of them, but I wouldn't feel comfortable with that.

Since we're already in "update script" territory we could also ask the upstream API about what the *BSR commit* was built from.
Now the API there seems to be protobuf over gRPC only, which makes it a bit unwieldy in a shell script (compared to god ol' curl&jq), however *buf* comes with a curl subcommand which allows us to get JSON from that API relatively cleanly.
The only part of the API which is public yet is the reflection API which is entirely irrelevant to us.
However since this API is used by *buf*, we can just use `--debug` and have a look.
Easy enough we get access to `buf.build/buf.registry.module.v1.CommitService/GetCommits` and can fetch [*BSR commit* information which includes [a source control reference](https://buf.build/bufbuild/registry/docs/main:buf.registry.module.v1#buf.registry.module.v1.Commit).
Well, an optional one, which isn't present for the one dependency I tested with, so that's a dead end.

### fetching the source

So if we went that way, a mix of hardcoding and guessing URLs it is?

So if we assume we get all the sources in a JSON file in the nixpkgs repo that can be updated with an update script we're set to pull those dependencies and build them.
The easy YAML path would be to include the git repo in the sources from where it pulls, but that leads us back to *buf* accessing the network.
Fetching them outside and passing them into the build as paths it is.
This is probably possible with some extensive yq-go hackery.

### building the source

Assuming we have a fully local *buf.yaml*, we can start building.
Or can we?
Well, given that zitadel only has three deps which AFAIK are all standalone with no further deps, that'd work out, but as soon as that changes we're met with recursive dependency resolution.
So let's assume we're fine, either because we implemented the recursive resolution in the update script or because we assert no further dependencies.
Now we also have to assume that *buf* can magically handle those repos.
Without the sources we've been fetching the already built protobuf files which have a very streamlined format since they're built for consumption by *buf*.
Now we might be building with different build systems, at which point this becomes rather convoluted and complex, but potentially possible.

Now this section alone has significant assumptions and required effort so it's unlikely we'll find maintainers willing to build and maintain that, as we'd basically vendor protobuf outside the source repo.

## skipping the build

### zitadel on buf.build

There's another option, technically.
We could fetch the entire zitadel protobuf build from *buf.build*; it's on there.
If we somehow managed to fetch the entirety of it including the dependencies, or maybe even go all-in and download the code-gen'd bindings from there (which seems to be a thing?) we'd only have to fetch that and FOD it.
I don't particularly like the idea, but if it could be as simple as fetching a tar and extracting it over the source repo in preBuild then I guess it works.
With the *BSR commit* being immutable and pinned this should work fine, so we'd just have to update that with respects to the corresponding zitadel version, which may or may not cause headaches.
Since the output should be tied to the commit we should also be able to fetch that using *buf* if need be, and not worry about version changes.
However I have not looked into that.… basically at all.

# Context

## BSR commit

A repository on a *BSR* has a "commit" structure for the artifacts like git has for repository contents.
Commits can be pinned in a *buf.lock* file.
They look like your usual sha256 and are a secure hash over the entire commit contents, so they are verifiable.
Since we're talking about the hash of the output, this has no 1:1 relation with the source code repository; source changes which do not touch the proto files or do not cause a semantic change do not generate a new commit.

## buf.build

*buf.build* is the flagship instance of a *BSR*.

## buf.yaml

The configuration file for a *buf* build.
Notably there are two versions, v1 and v2, with a different structure and slightly different feature set.
The v2 is also much closer to the buf workspace configuration for v1, but differences exist.

## buf.lock

Similar to Cargo.lock or whatever else system you know, *buf* allows pinning a dependency to a specific *BSR commit*, thereby allowing you to reproduce a build (if you have connectivity of course).

## buf

*buf* is a CLI tool for building protobuf specs and using plugins to generate bindings.

## BSR

A Buf Schema Registry or *BSR* is sort of a combined CI service which gets protobuf input using *buf*, and builds it, offers artifacts, documentation, and some extra features.
This is how *buf* builds plugins since plugins are basically containers and [buf does not want to depend on a container runtime](https://github.com/bufbuild/buf/issues/2543#issuecomment-1841804678).

It also offers a *BSR API*.

The *BSR API* is a gRPC and protobuf based API offered by *BSR* instances for querying information.
As far as I can tell there is no REST API available?

Signed-off-by: benaryorg <binary@benary.org>

benaryorg committed:
pkgs/zitadel: buf resiliency & update script

Below is a write-up on the current state of covering *buf* in nixpkgs, in relation to the zitadel package which was intended for a comment in a PR originally, however since this commit implements one of the approaches we may as well discuss it as part of the PR.
Please note that I am not an expert in zitadel, protobuf, golang, Nix, and especially not buf.build or buf in general, so many grains of salt adviced.

This solution has opted for *pointing to buf export data*, which requires that a bunch of hashes from FODs get precomputed.
Because the FOD is a derivation (three of them actually) with a command that actually runs, this is not trivial.
However the script also covers the repo itself by checking GitHub for updates, yarn, the protobuf dependencies, and since it is the only FOD hash left also the go modules.
The repo and yarn work with the usual pre-fetchers.
Protobuf by running the buf commands with modified cache dir (so as to not bother anyone with extra files lying around).
And the go modules, by virtue of being the only thing left, can be resolved by trying to build the package with lib.fakeHash and getting the FOD error and parsing that.

buf building
============

Let's start by describing the problem, our prior approach, and its issues.
The problem is relatively simple, we need to build the protobuf specs and get bindings for them, however those specs depend on other specs, and the only reference we have is to an online service which gives us artifacts.
So since we don't have the source repositories for any of this we have to make do by fetching the artifacts in an FOD (to allow us network access for fetching it) using upstream tooling.
The approach has the issue of the FOD being dependent on the upstream tooling which on version change occasionally changes its output and therefore the FOD hash.

Based on this a change can mean several possible things, which can roughly be classified by how the build is changed.
Since many of those approaches have obstacles that are the same for each subset there is a context section further down so I don't have to explain the same thing multiple times.
Terms described in the context have *emphasis*.

same build, extending the fetch
-------------------------------

fetchtarball
^^^^^^^^^^^^

Since the dependency is based on the *BSR commit* we could include that commit in a fetcher much like we include git revs.
If we had a deterministic fetcher which could fetch the *BSR commit* (or rather its output), then we would not depend on *buf*, and therefore not be affected by its version changes as part of the FOD.
The build process would then be separate but offline, thus not an FOD, so a version change in *buf* would also cause a rebuild of the bindings and package.

However *buf.build* has apparently [disabled (or broken) their tar export with the examples on their docs not working](https://buf.build/docs/bsr/module/export/#export-with-curl) as far as I can tell.
This means we cannot use a tarball fetcher.

BUF_CACHE_DIR
^^^^^^^^^^^^^

By keeping the build exactly the same but running part of it (`buf dep graph`) beforehand to fetch all dependencies, FODing the cache, and reinjecting it into the build, we can remove the FOD from the actual build.
However testing *nixos-24.11* vs. *master* nixpkgs *buf*, the cache dir shows differences (newer version creates an empty folder for plugins), which means we're back to the point where we depend on *buf* versions for the output of an *FOD*, which on *buf* updates can cause hash mismatches.
The only difference is that now we're hashing internals instead of output-related data, which might prove more stable, but can also change without notice (e.g. in a patch version) so even small updates can cause this.
Prebuilding the cache dir ourselves doesn't seem quite feasible since it includes a plethora of extra files including YAML files generated from build metadata.

naive buf export
^^^^^^^^^^^^^^^^

*buf* has an export subcommand which fetches the build artifacts and outputs them in an output directory.
Since this includes only the built output of what is part of the secure hash of the *BSR commit*, it will most likely not change in between *buf* versions.
The problem here lies with integrating it into the build since we cannot point *buf* at the directory and say "go" without modifying the build files significantly.

changing the build
------------------

pointing to buf export data
^^^^^^^^^^^^^^^^^^^^^^^^^^^

If we used the naive *buf* export approach from above, we'd have to somehow point *buf* at the new data.
Unfortunately the *buf.yaml* is structured in a way that we cannot easily replace those paths since zitadel uses the v1 version and the "deps" key which only accepts a *BSR* reference and nothing else (no path, no git repo, no nothing).
Moving the import of that data elsewhere would require changes to the yaml files which is quite hacky tbh.
Especially since we're now pulling in already built protobuf files vs. the original source repositories.
Luckily protobuf itself has an import system which accomodates files lying around in directories, that's how it usually works without *buf*.
By simply copying the prefetched directories over and erasing the deps from the *buf.yaml* as well as *buf.lock* and instead adding them to the buf.work.yaml which lists the project-wide directories to be built we can shorten out every *buf* logic pertaining to those files and treat them as natural protobuf files.
It still involves wild hacking around in *buf* configuration, however the code pulls from the repository itself.
As long as upstream doesn't introduce any indirect dependencies this should continue to work indefinitely.

source code all the way down
----------------------------

Instead of fetching the build artifacts, which are the thing fetched online, we could try and fetch the sources of the dependencies.
This has multiple issues explored in the next few sections.

getting the source repository
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

First of all we would have to get the source repositories, in case upstream ever changes these.
These are technically easily guessable (replace *buf.build* with github.com works for most of them, but I wouldn't feel comfortable with that.

Since we're already in "update script" territory we could also ask the upstream API about what the *BSR commit* was built from.
Now the API there seems to be protobuf over gRPC only, which makes it a bit unwieldy in a shell script (compared to god ol' curl&jq), however *buf* comes with a curl subcommand which allows us to get JSON from that API relatively cleanly.
The only part of the API which is public yet is the reflection API which is entirely irrelevant to us.
However since this API is used by *buf*, we can just use `--debug` and have a look.
Easy enough we get access to `buf.build/buf.registry.module.v1.CommitService/GetCommits` and can fetch [*BSR commit* information which includes [a source control reference](https://buf.build/bufbuild/registry/docs/main:buf.registry.module.v1#buf.registry.module.v1.Commit).
Well, an optional one, which isn't present for the one dependency I tested with, so that's a dead end.

fetching the source
^^^^^^^^^^^^^^^^^^^

So if we went that way, a mix of hardcoding and guessing URLs it is?

So if we assume we get all the sources in a JSON file in the nixpkgs repo that can be updated with an update script we're set to pull those dependencies and build them.
The easy YAML path would be to include the git repo in the sources from where it pulls, but that leads us back to *buf* accessing the network.
Fetching them outside and passing them into the build as paths it is.
This is probably possible with some extensive yq-go hackery.

building the source
^^^^^^^^^^^^^^^^^^^

Assuming we have a fully local *buf.yaml*, we can start building.
Or can we?
Well, given that zitadel only has three deps which AFAIK are all standalone with no further deps, that'd work out, but as soon as that changes we're met with recursive dependency resolution.
So let's assume we're fine, either because we implemented the recursive resolution in the update script or because we assert no further dependencies.
Now we also have to assume that *buf* can magically handle those repos.
Without the sources we've been fetching the already built protobuf files which have a very streamlined format since they're built for consumption by *buf*.
Now we might be building with different build systems, at which point this becomes rather convoluted and complex, but potentially possible.

Now this section alone has significant assumptions and required effort so it's unlikely we'll find maintainers willing to build and maintain that, as we'd basically vendor protobuf outside the source repo.

skipping the build
------------------

zitadel on buf.build
^^^^^^^^^^^^^^^^^^^^

There's another option, technically.
We could fetch the entire zitadel protobuf build from *buf.build*; it's on there.
If we somehow managed to fetch the entirety of it including the dependencies, or maybe even go all-in and download the code-gen'd bindings from there (which seems to be a thing?) we'd only have to fetch that and FOD it.
I don't particularly like the idea, but if it could be as simple as fetching a tar and extracting it over the source repo in preBuild then I guess it works.
With the *BSR commit* being immutable and pinned this should work fine, so we'd just have to update that with respects to the corresponding zitadel version, which may or may not cause headaches.
Since the output should be tied to the commit we should also be able to fetch that using *buf* if need be, and not worry about version changes.
However I have not looked into that.… basically at all.

Context
=======

BSR commit
----------

A repository on a *BSR* has a "commit" structure for the artifacts like git has for repository contents.
Commits can be pinned in a *buf.lock* file.
They look like your usual sha256 and are a secure hash over the entire commit contents, so they are verifiable.
Since we're talking about the hash of the output, this has no 1:1 relation with the source code repository; source changes which do not touch the proto files or do not cause a semantic change do not generate a new commit.

buf.build
---------

*buf.build* is the flagship instance of a *BSR*.

buf.yaml
--------

The configuration file for a *buf* build.
Notably there are two versions, v1 and v2, with a different structure and slightly different feature set.
The v2 is also much closer to the buf workspace configuration for v1, but differences exist.

buf.lock
--------

Similar to Cargo.lock or whatever else system you know, *buf* allows pinning a dependency to a specific *BSR commit*, thereby allowing you to reproduce a build (if you have connectivity of course).

buf
---

*buf* is a CLI tool for building protobuf specs and using plugins to generate bindings.

BSR
---

A Buf Schema Registry or *BSR* is sort of a combined CI service which gets protobuf input using *buf*, and builds it, offers artifacts, documentation, and some extra features.
This is how *buf* builds plugins since plugins are basically containers and [buf does not want to depend on a container runtime](https://redirect.github.com/bufbuild/buf/issues/2543#issuecomment-1841804678).

It also offers a *BSR API*.

The *BSR API* is a gRPC and protobuf based API offered by *BSR* instances for querying information.
As far as I can tell there is no REST API available?

Signed-off-by: benaryorg <binary@benary.org>

nrabulinski commented:
I'm sorry, this got completely buried and I forgot to get back to this. Could you please reformat and rebase on latest master as I review? Assuming you still want to spend brain cycles on this. I won't fault you if you don't

benaryorg commented:
> I'm sorry, this got completely buried and I forgot to get back to this.

No problem. I just rebased and.… oh crap, just realized I probably downgraded zitadel in the course of doing so.…
I'll update this to the version it was at currently so we have no change in version and if upstream is ahead we should be getting a PR for an update to see whether it works too.

> Could you please reformat and rebase on latest master as I review? Assuming you still want to spend brain cycles on this. I won't fault you if you don't

Don't worry, I don't mind, I'm used to rebasing and merge conflict resolution ^^
I'm more irked about having just run `treefmt` on the entire repo just to have exactly zero format changes, so the Nix files are fine.
Did you mean the Nix files or Python specifically, if it's Python I wouldn't mind a pointer in the right direction about where to find the formatting rules/formatter for that.

benaryorg commented:
Good news, upstream is not ahead and we get the most up to date version right away (also good news, the update script did its job just fine for that).
Not-so-good news, since nixpkgs *master* is quite a bit ahead of the package cache I'll have to keep compiling for a bit longer to see if this actually works (building gcc is fine, but it'll take a while).

I've pushed the changes anyway while my local machine is compiling.

nrabulinski commented on pkgs/by-name/zi/zitadel/package.nix:
```diff
@@ -72,20 +101,31 @@ let
+        yq --inplace '.deps = []' proto/buf.yaml
+        yq --inplace '.deps = []' proto/buf.lock
+        yq eval-all --inplace '[ select(fileIndex == 0),select(fileIndex == 1) ] | .[0].directories += .[1] | .[0]' buf.work.yaml - <<"END"
+        ${builtins.toJSON (builtins.attrNames generatedSrc.protobufDeps)}
```
I'd rather we used `passAsFile` here

benaryorg replied:
Oh, I somehow remember trying to avoid passing multiple files and instead using stdin for *yq-go* because – compared to *jq* – it seems to have a more permissive way of providing arguments and I wanted to avoid it being "smart", but since I didn't get around the `-` anyway I'll use *passAsFile*, no problem.

benaryorg replied:
Small addendum: I did it specifically because I didn't know how `--inplace` and passing the additional file would interact.

Either way I just tried applying this:

```diff
diff --git a/pkgs/by-name/zi/zitadel/package.nix b/pkgs/by-name/zi/zitadel/package.nix
index 41b2e6e75..29eece5b1 100644
--- a/pkgs/by-name/zi/zitadel/package.nix
+++ b/pkgs/by-name/zi/zitadel/package.nix
@@ -104,14 +104,15 @@ let
         yq-go
       ];

+      protobufReplacements = builtins.toJSON (builtins.attrNames generatedSrc.protobufDeps);
+      passAsFile = [ "protobufReplacements" ];
+
       buildPhase = ''
         runHook preBuild

         yq --inplace '.deps = []' proto/buf.yaml
         yq --inplace '.deps = []' proto/buf.lock
-        yq eval-all --inplace '[ select(fileIndex == 0),select(fileIndex == 1) ] | .[0].directories += .[1] | .[0]' buf.work.yaml - <<"END"
-        ${builtins.toJSON (builtins.attrNames generatedSrc.protobufDeps)}
-        END
+        yq eval-all --inplace '[ select(fileIndex == 0),select(fileIndex == 1) ] | .[0].directories += .[1] | .[0]' buf.work.yaml $protobufReplacements
         ${lib.concatLines (
           lib.flip lib.mapAttrsToList generatedSrc.protobufDeps (
             name: dep: "cp -r ${fetchProtobufDep dep} ${name}"
```

and it bailed with:

```text
DEBUG   buffetch termination found      {"curDirPath": "build/source", "path": "build/source"}
DEBUG   buffetch termination found      {"curDirPath": ".", "path": "."}
DEBUG   github.com/bufbuild/buf/private/buf/bufworkspace.(*workspaceProvider).getWorkspaceForBucketAndModuleDirPathsV1Beta1OrV1 {"duration": "5.26µs"}
DEBUG   github.com/bufbuild/buf/private/buf/bufctl.(*controller).getWorkspaceForSourceRef       {"duration": "577.522µs"}
DEBUG   github.com/bufbuild/buf/private/bufpkg/bufimage.BuildImage      {"duration": "60.447563ms"}
proto/zitadel/object.proto:4:8:import "protoc-gen-openapiv2/options/annotations.proto": file does not exist
```

Unless I somehow messed up the patch, I don't think this will work.

nrabulinski replied:
Hmmm... I'm not sure why something like `yq eval-all --inplace '[ select(fileIndex == 0),select(fileIndex == 1) ] | .[0].directories += .[1] | .[0]' buf.work.yaml - < $protobufReplacements` wouldn't work? That should produce exactly the same result while not requiring the heredoc hack

benaryorg replied:
Darn, GitHub only refreshed the rest of the PR, not the discussion.

Fast forward half an hour of debugging, give or take.… `$protobufReplacements` vs. `$protobufReplacementsPath`

It builds just fine and I am tempted to correct some of the stuff I found in nixpkgs that I'm not sure how it ever worked.
Specifically [the akkoma override howto](https://github.com/NixOS/nixpkgs/blob/1a11438b83b5694ed1a4e2b950ed7c111f6fe1cf/nixos/modules/services/web-apps/akkoma.md?plain=1#L194) and the random derivation I opened yesterday to double check my syntax.

Resolving this.

nrabulinski commented on pkgs/by-name/zi/zitadel/package.nix:
```diff
@@ -72,20 +101,31 @@ let
+        yq eval-all --inplace '[ select(fileIndex == 0),select(fileIndex == 1) ] | .[0].directories += .[1] | .[0]' buf.work.yaml - <<"END"
+        ${builtins.toJSON (builtins.attrNames generatedSrc.protobufDeps)}
+        END
+        ${lib.concatLines (
+          lib.flip lib.mapAttrsToList generatedSrc.protobufDeps (
+            name: dep: "cp -r ${fetchProtobufDep dep} ${name}"
+          )
+        )}
```
This is clever piece of code, but could we make all protobuf deps into a single `symlinkJoin` (or `linkFarm` if you want to rename them)? And then simply copy everything from that into the builder. Also, symlinks are preferable, but I assume zitadel doesn't like that?

benaryorg replied:
Issue is that this needs to be inside the source tree, which means (unless I'm mistaken about how it works) *symlinkJoin* and *linkFarm* won't help much.
The former because it's built to merge output trees and it would need to happen as part of passing *src*, and therefore would make the `yq --inplace` work on symlinks and I'm not sure how well that'd work (it might try to write to the target which'd err out with EPERM).
The latter would just create a tree which'd still need to be merged into *src* which again sounds tricky.

I'll see if it still builds with symlinks since that's a cheap `ln -s` rather than `cp -r` which might speed up builds.

(if I'm mistaken about anything above *please* tell me ^^)

benaryorg replied:
The `ln -s` version builds nicely, so I'll resolve this, feel free to reopen though if you can think of a different approach that I failed to see (especially in regards to *symlinkJoin* and *linkFarm*.

nrabulinski replied:
My idea was to do something like this
```nix
let
  ...
  protobufDeps = linkFarm "zitadel-protobuf-deps" (lib.mapAttrsToList (name: dep: { inherit name; path = fetchProtobufDep dep; }) generatedSrc.protobufDeps);
  ...
  cp -r "${protobufDeps}"/* ./
```
So we don't have to end up generating bash. But it's not the end of the world and if you don't like that I'd at least more like the inner expression to be extracted into a variable so the code to parse is less dense

nrabulinski reviewed (changes requested):
Other than that LGTM.
Re: formatting - no, I only meant the Nix files. I mentioned it because of the global reformat so a lot of the PRs had to be reformatted :)

benaryorg commented:
> Other than that LGTM.

Thanks for the input, much appreciated.
I changed the symlink thing and resolved the thread (again, feel free to reopen).
The other one I double checked my syntax and I'm reasonably sure I didn't mess up, but I didn't feel sure enough to resolve that one, so if you don't see anything wrong with my patch please resolve that one yourself (unless you have further input of course).

Otherwise, merge at your leisure.?
Since we were already on the most current version we won't get to see the script in action until the next release though.
Please ping me when that happens, I'd like to know that it worked ^^

(also the offer to add me as a maintainer still stands, though I did move over to Authentik so I'm not an ideal maintainer, but better than none at all I guess)

nrabulinski commented:
> (also the offer to add me as a maintainer still stands, though I did move over to Authentik so I'm not an ideal maintainer, but better than none at all I guess)

I'll gladly take you up on the offer, I haven't migrated myself yet so I still have a horse in this race

benaryorg committed:
zitadel: Remove `with lib`

as per https://redirect.github.com/NixOS/nixpkgs/issues/371862

Personally I like the `builtins.attrValues { inherit (lib.maintainers) maintainer1 maintainer 2; }` trick to avoid `with` for lists too, YMMV of course.

Signed-off-by: benaryorg <binary@benary.org>

benaryorg committed:
zitadel: Add benaryorg as Co-maintainer

Signed-off-by: benaryorg <binary@benary.org>

benaryorg commented:
> I'll gladly take you up on the offer

Alright, there we go.

benaryorg commented:
A little bit delayed I randomly figured out why the *passAsFiles* didn't work, and once again I'm sorry for the extraneous pushes (both for you as the reviewer as well as the testing infrastructure which is building zitadel for the sixth time or so I guess).

*/nix/store/q5hnrq5vw8zm6678zaqjvh648sbva2il-zitadel-2.71.7* has compiled properly and it looks like it works, and given that it compiles at all suffices in the context of this PR for my tastes.
For anything else a NixOS Test might be nice, but testing anything beyond "does it start" sounds like a *lot* of REST API usage which.… maybe some time, but not today.…

benaryorg commented:
FWIW I've been thinking about how to handle NixOS stable, and it [doesn't look like there's a flag](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#automatic-package-updates) that passes the "branch" the script is acting on, otherwise I would've taken a stab at making it so the package always follows the newest release on *master*, but sticks to the current minor version on the stable branches (if that's even a thing with that slice of update infrastructure, doesn't quite look like it).

benaryorg commented:
@nrabulinski I've rebased this onto *master* and checked whether it still works (it does, running the script will update Zitadel from 2.71.7 to 3.2.0).
Is there anything still blocking this (or any changes you'd like to see before merging)?

Otherwise my suggestion would be to merge this, wait for the bot-created update PR (to 3.2.0), and then handle everything else (primarily checking whether the new version still *works*, at least I can confirm that it *builds* as of now) over in that PR.?

Oh, no worries about any delays by the way, I don't mind.
I just want to make sure that this isn't blocked on my end right now.

nrabulinski reviewed (approved):
I'm really sorry, I completely forgot to get back to you. The code and changes look good, and everything seems to eval and work correctly. Thank you so much for all the hard work and sticking with this PR!

benaryorg commented:
> I'm really sorry, I completely forgot to get back to you.

Nothing to worry about, thanks for approving.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2425

r-vdp commented:
@benaryorg are you still up to maintain this going forward?  If so, I'll merge once CI passed, thanks for all your work on this!

(Close/reopen combo was to retrigger CI because of recent changes to required status checks for the repo.)
